pydantic/_internal/_decorators.py

```diff
@@ -341,66 +349,93 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
 _FUNCS: set[str] = set()
 
 
-def prepare_serializer_decorator(
-    function: Callable[..., Any] | classmethod[Any] | staticmethod[Any], allow_reuse: bool
-) -> Callable[..., Any] | classmethod[Any]:
+AnyDecoratorCallable: TypeAlias = 'Union[ classmethod[Any], staticmethod[Any], partialmethod[Any], Callable[..., Any]]'
+
+
+def unwrap_wrapped_function(
+    func: Any,
+    *,
+    unwrap_class_static_method: bool = True,
+) -> Any:
+    all: tuple[Any, ...]
+    if unwrap_class_static_method:
+        all = (
+            staticmethod,
+            classmethod,
+            partial,
+            partialmethod,
+        )
+    else:
+        all = (
+            partial,
+            partialmethod,
+        )
+
+    while isinstance(func, all):
+        if unwrap_class_static_method and isinstance(func, (classmethod, staticmethod)):
+            func = func.__func__
+        elif isinstance(func, (partial, partialmethod)):
+            func = func.func
+
+    return func
+
+
+def get_function_ref(func: Any) -> str:
+    func = unwrap_wrapped_function(func)
+    return (
+        getattr(func, '__module__', '<No __module__>')
+        + '.'
+        + getattr(func, '__qualname__', f'<No __qualname__: id:{id(func)}>')
+    )
+
+
+def prepare_serializer_decorator(function: AnyDecoratorCallable, allow_reuse: bool) -> Any:
     """
     Warn about validators/serializers with duplicated names since without this, they can be overwritten silently
     which generally isn't the intended behaviour, don't run in ipython (see #312) or if `allow_reuse` is True.
     """
-    if isinstance(function, staticmethod):
-        function = function.__func__
     if not allow_reuse and not in_ipython():
-        ref = f'{function.__module__}::{function.__qualname__}'
+        ref = get_function_ref(function)
```

Comment:
In `check_for_duplicate_decorator_function ` below you do:
```python
        function = unwrap_wrapped_function(function)
        ref = get_function_ref(function)
```
Whereas here, you don't do the unwrapping. Is this intentional? Without digging into the full logic path it seems like it should be the same in both places but I could imagine it being possible you already know the function has been unwrapped here, or similar.

Reply:
I just deleted this function _and_ deleted the `function = unwrap_wrapped_function(function)` in `check_for_duplicate_decorator_function` because `get_function_ref` already does that.